cpp/src/quantiles/quantiles.cu

```diff
@@ -61,48 +60,216 @@ quantiles(table_view const& input,
                           mr);
 }
 
-}
+template<typename DataIterator, typename Result>
+struct quantiles_data_functor
+{
+    DataIterator data;
+    size_type size;
+    interpolation interp;
+
+    Result __device__ operator()(double q) {
+        return detail::select_quantile_data<Result>(data, size, q, interp);
+    }
+};
+
+template<typename ValidityIterator>
+struct quantiles_mask_functor
+{
+    ValidityIterator validity;
+    size_type size;
+    interpolation interp;
+
+    bool __device__  operator() (double q) {
+        return detail::select_quantile_validity(validity, size, q, interp);
+    }
+};
+
+
+template<typename SortMapIterator>
+struct quantiles_functor
+{
+
+    SortMapIterator ordered_indices;
+    std::vector<double> const& q;
+    interpolation interp;
+    bool retain_types;
+    rmm::mr::device_memory_resource* mr;
+    cudaStream_t stream;
+
+    template<typename T, typename... Args>
+    std::enable_if_t<not std::is_arithmetic<T>::value, std::unique_ptr<column>>
+    operator()(Args&&... args)
+    {
+        CUDF_FAIL("Only arithmetic types are supported in quantiles.");
+    }
+
+    template<typename T>
+    std::enable_if_t<std::is_arithmetic<T>::value, std::unique_ptr<column>>
+    operator()(column_view const& input)
+    {
+        auto type = retain_types
+            ? input.type()
+            : data_type{type_to_id<double>()};
+
+        auto output = make_fixed_width_column(type,
+                                              q.size(),
+                                              mask_state::UNALLOCATED,
+                                              stream,
+                                              mr);
+
+        if (output->size() == 0) {
+            return output;
+        }
+
+        auto d_input = column_device_view::create(input);
+        auto d_output = mutable_column_device_view::create(output->mutable_view());
+
+        rmm::device_vector<double> q_device{q};
+
+        auto sorted_data = thrust::make_permutation_iterator(input.data<T>(),
+                                                             ordered_indices);
 
+        if (retain_types)
+        {
+            thrust::transform(
+                q_device.begin(),
+                q_device.end(),
+                d_output->template begin<T>(),
+                [sorted_data, interp=interp, size=input.size()]
+                __device__ (double q){
+                    return select_quantile_data<T>(sorted_data, size, q, interp);
```

Comment:
I'm not not sure why the `retain_types` logic needs to be inside `quantiles`, but even if it is, I think the side effects would be much clearer to move the cast outside of `select_quantile_data`. This would make `select_quantile_data` more in line with "Do One Thing" also. 
```suggestion
                    return static_cast<T>(select_quantile_data(sorted_data, size, q, interp));
```

Reply:
Since quantiles is a mathematical operation, the operation cannot be ignorant of the underlying type. From a design/SOLID perspective, I completely agree with you, but unfortunately I think this is one of those times a concession needs to be made in order to maintain precision. :/